core-bundle/src/Security/Voter/DataContainer/AbstractDynamicPtableVoter.php

```diff
@@ -78,16 +78,33 @@ private function getParentTableAndId(int $id): array
 
         // Limit to a nesting level of 10
         $records = $this->connection->fetchAllAssociative(
-            "SELECT id, @pid:=pid AS pid, ptable FROM $table WHERE id=?".str_repeat(" UNION SELECT id, @pid:=pid AS pid, ptable FROM $table WHERE id=@pid", 9),
-            [$id],
+            "SELECT id, @pid:=pid AS pid, ptable FROM $table WHERE id=?".str_repeat(" UNION SELECT id, @pid:=pid AS pid, ptable FROM $table WHERE id=@pid AND ptable=?", 9),
+            [$id, ...array_fill(0, 9, $table)],
         );
 
+        if (!$records) {
+            throw new \RuntimeException(\sprintf('Parent record of %s.%s not found', $table, $id));
+        }
+
+        $last = end($records);
+
+        // If the given $id is the child of a record where ptable!=$table (e.g. only one
+        // element nested), $records will only have one result, and we can directly use it.
+        if ($last['ptable'] !== $table) {
+            return $this->parents[$id] = $last;
+        }
+
         // Trigger recursion in case our query returned exactly 10 records in which case
         // we might have higher parent records
         if (10 === \count($records)) {
-            $records = array_merge($records, $this->getParentTableAndId((int) end($records)['pid']));
+            return $this->getParentTableAndId((int) $last['pid']);
```

Comment:
IMHO recursion on a function level should be avoided as much as possible, so that we can switch to/dynamically use a non-recursive version (like `WITH RECURSIVE` for MySQL 8.0+) later on. The recursion - for now - could also happen inside the function with an anonymous function. 

As briefly discussed on Slack, the logic could also be put into a utility class as it is also used in `Database::getChildRecords()`.

Just noting these two things. We can add or change bot of this in a followup PR.

Reply:
Yes, that is planned for Contao 5.5 already (see #6682).